### PR TITLE
cmd/update-report: improve formula file validation

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Currently, ruby files that are not in the Cask directory are considered to be formulae which doesn't make sense. We know that these should only be in a few directories so we can check for that explicitly.

Beyond that the `Tap#cask_file?` and `Tap.formula_file?` methods were only used inside update-report so it doesn't make sense to turn them into pathnames and expand things when we know that each string will be a relative path from a tap that we can just check with a regex.

This change will stop other tap changes like new commands or changes to other directories from showing up as new formulae.

### CODE REVIEW

File: Library/Homebrew/cmd/update-report.rb [hunk 1]
```diff
@@ -344,6 +344,10 @@ def migrate_gcc_dependents_if_needed
 require "extend/os/cmd/update-report"
 
 class Reporter
+  # Validates the relative path from the tap directory returned by `git diff`.
+  FORMULA_PATH_REGEX = %r{^((Homebrew)?Formula/([^/]+/)*)?[^/]+\.rb}.freeze
```
Comment (apainintheneck) [inline Library/Homebrew/cmd/update-report.rb:348]: Corresponds to this directory list.

https://github.com/Homebrew/brew/blob/e96fd24a7328a4913de2f7c56fd2c97d9aaa4572/Library/Homebrew/tap.rb#L468-L471

File: Library/Homebrew/tap.rb [hunk 2]
```diff
@@ -566,31 +566,6 @@ def ruby_file?(file)
...
-  # @private
-  sig { params(file: T.any(String, Pathname)).returns(T::Boolean) }
-  def cask_file?(file)
-    file = Pathname.new(file) unless file.is_a? Pathname
-    file = file.expand_path(path)
-    return false unless ruby_file?(file)
-
-    file.to_s.start_with?("#{cask_dir}/")
-  end
-
```
Comment (apainintheneck) [inline Library/Homebrew/tap.rb:593]: These were only used in the update report and `Tap#formula_files` above which has been updated.

Comment (reitermarkus) [reply Library/Homebrew/tap.rb:593]: `cask_file?` is used for cask CI currently. There it is also used in combination with `git diff`, so may be best to move the `git diff` logic into `Tap`, so it can be used to get all changed files between two commits.

Comment (apainintheneck) [reply Library/Homebrew/tap.rb:593]: Could you point out where specifically it's used in cask CI? I couldn't find it with a quick grep of the cask repo.

Both of these were marked `@private` so I didn't check other repos but now that you mentioned it I looked and one of them is used in the test bot repo.

```
/u/l/H/L/T/h/homebrew-test-bot (master|✔) $ git grep -F '.cask_file?'
/u/l/H/L/T/h/homebrew-test-bot (master|✔) [1]$ git grep -F '.formula_file?'
lib/tests/formulae_detect.rb:          next unless tap.formula_file?(file)
```

File: Library/Homebrew/cmd/update-report.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/cmd/update-report.rb:348]: This regex doesn't work with the way we process formula directories internally. It should be updated to use the first valid directory of this list or maybe there is a better way to check for formula.

File: Library/Homebrew/tap.rb (hunk 2 above)
Comment (apainintheneck) [reply Library/Homebrew/tap.rb]: I just rebased this to add back in the `.cask_file?` and `.formula_file?` methods because they are used in other repos. I made both of them more restrictive though. The use of `.formula_file?` in test bot would need to be changed first though but the current method already takes a string so that shouldn't be too hard.

The idea of combining the `git diff` logic is an interesting idea. I'd have to look to see how much logic is shared there between different use cases.

File: Library/Homebrew/tap.rb [hunk 3]
```diff
@@ -566,29 +568,35 @@ def ruby_file?(file)
...
+  # @private
+  sig { returns(Regexp) }
+  def formula_file_regex
+    @formula_file_regex ||= case formula_dir.basename.to_s
+    when "Formula"
+      if official?
+        %r{^Formula(/[^/]+)+\.rb$}
+      else
+        %r{^Formula/[^/]+\.rb$}
+      end
```
Comment (apainintheneck) [inline Library/Homebrew/tap.rb:594]: Not sure if this difference is actually meaningly in most cases.

**Review by apainintheneck — COMMENTED** [review]
Note to self: It'd probably be a good idea to add a test for casks in the update report.

File: Library/Homebrew/tap.rb (hunk 3 above)
Comment (reitermarkus) [reply Library/Homebrew/tap.rb:594]: Does that mean if e.g. `Formula/` exists, any Ruby files in the root are not considered formula files?

**Comment by MikeMcQuaid** [comment]
> Beyond that the Tap#cask_file? and Tap.formula_file? methods were only used inside update-report so it doesn't make sense to turn them into pathnames and expand things when we know that each string will be a relative path from a tap that we can just check with a regex.

They are also used in https://github.com/homebrew/homebrew-test-bot/

File: Library/Homebrew/tap.rb [hunk 4]
```diff
@@ -566,29 +568,35 @@ def ruby_file?(file)
...
+    when "HomebrewFormula"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:595]: IMO we should consider deprecating both `HomebrewFormula` and having formula files in the root of a tap.

File: Library/Homebrew/test/tap_spec.rb [hunk 5]
```diff
@@ -173,7 +173,6 @@ def setup_completion(link:)
     expect(homebrew_foo_tap.tap_migrations).to eq("removed-formula" => "homebrew/foo")
     expect(homebrew_foo_tap.command_files).to eq([cmd_file])
     expect(homebrew_foo_tap.to_hash).to be_a(Hash)
-    expect(homebrew_foo_tap).to have_formula_file(formula_file)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/test/tap_spec.rb:176]: Why this change?
